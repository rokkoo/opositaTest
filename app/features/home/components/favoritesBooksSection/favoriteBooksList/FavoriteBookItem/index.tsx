import React from 'react';
import { StyleSheet, View } from 'react-native';
import Icon from 'react-native-vector-icons/AntDesign';

import BookItem, {
  BookItemProps,
} from '@app/features/home/components/bookItem';
import { Spacing } from '@app/theme/metric';
import { Colors } from '@app/theme/colors';

interface FavoriteBookItemProps extends BookItemProps {}

const FavoriteBookItem: React.FC<FavoriteBookItemProps> = props => {
  return (
    <View testID="book-item">
      <BookItem {...props} />
      <View style={styles.container}>
        <Icon name={'star'} size={24} color={Colors.yellow} />
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    right: Spacing.m,
    top: Spacing.m,
  },
});

export default React.memo(FavoriteBookItem);
