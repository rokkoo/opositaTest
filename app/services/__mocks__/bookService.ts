jest.mock('../booksService', () => {
  return {
    BooksService: {
      getBooks: jest.fn().mockResolvedValue([
        {
          url: 'https://www.anapioficeandfire.com/api/books/1',
          name: 'A Game of Thrones',
          isbn: '978-0553103540',
          authors: ['George R. R. Martin'],
          numberOfPages: 694,
          publisher: 'Bantam Books',
          country: 'United States',
          mediaType: 'Hardcover',
          released: '1996-08-01T00:00:00',
          characters: ['https://www.anapioficeandfire.com/api/characters/2'],
          povCharacters: [
            'https://www.anapioficeandfire.com/api/characters/148',
          ],
        },
      ]),
    },
  };
});
