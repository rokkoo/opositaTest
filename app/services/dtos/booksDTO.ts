import { format } from 'date-fns';
import { es } from 'date-fns/locale';

export interface Book {
  readonly url: string;
  readonly name: string;
  readonly isbn: string;
  readonly authors: string[];
  readonly publisher: string;
  readonly country: string;
  readonly mediaType: string;
  readonly released: string;
  readonly characters: string[];
  readonly numberOfPages: number;
}

// https://anapioficeandfire.com/Documentation#books
export class BooksDTO {
  static getArray(data: any[]): Book[] {
    return data.map(book => {
      const date = new Date(book.released);
      const released = format(date, 'd MMMM, yyyy', {
        locale: es,
      });

      const dto: Book = {
        url: book.url,
        name: book.name,
        isbn: book.isbn,
        authors: book.authors,
        publisher: book.publisher,
        country: book.country,
        mediaType: book.mediaType,
        released: released,
        characters: book.characters,
        numberOfPages: book.numberOfPages,
      };

      return dto;
    });
  }
}
