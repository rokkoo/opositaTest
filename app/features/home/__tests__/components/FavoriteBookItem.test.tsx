import React from 'react';
import { screen, waitFor } from '@testing-library/react-native';

import { mockBooksData } from '@app/services/__mocks__/bookService';
import { BooksDTO } from '@app/services/dtos/booksDTO';
import { renderWithReactQuery } from '@app/helpers/tests/renderWithReactQuery';
import { AppTheProvider } from '@app/helpers/tests/appTheProvider';
import FavoriteBookItem from '../../components/favoritesBooksSection/favoriteBooksList/FavoriteBookItem';

const mockBook = BooksDTO.getArray(mockBooksData)[0];

describe('Favorite Book item', () => {
  test('it should render correctly', async () => {
    renderWithReactQuery(
      <AppTheProvider>
        <FavoriteBookItem {...mockBook} index={0} />
      </AppTheProvider>,
    );

    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('it should render book name', async () => {
    renderWithReactQuery(
      <AppTheProvider>
        <FavoriteBookItem {...mockBook} index={0} />
      </AppTheProvider>,
    );

    await waitFor(() => {
      const title = screen.queryByText(mockBook.name);
      expect(title).toBeTruthy();
    });
  });

  test('it should render book cover image', async () => {
    renderWithReactQuery(
      <AppTheProvider>
        <FavoriteBookItem {...mockBook} index={0} />
      </AppTheProvider>,
    );

    await waitFor(() => {
      const title = screen.queryByTestId(`img${mockBook.isbn}`);
      expect(title).toBeTruthy();
    });
  });

  test('it should render favorite icon', async () => {
    renderWithReactQuery(
      <AppTheProvider>
        <FavoriteBookItem {...mockBook} index={0} />
      </AppTheProvider>,
    );

    await waitFor(() => {
      const title = screen.queryByTestId('favoriteIcon');
      expect(title).toBeTruthy();
    });
  });
});
