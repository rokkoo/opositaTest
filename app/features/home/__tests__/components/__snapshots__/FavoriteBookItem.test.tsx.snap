// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorite Book item it should render correctly 1`] = `
<View
  testID="book-item"
>
  <View
    style={
      {
        "marginLeft": 0,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Image
        height={200}
        onError={[Function]}
        source={
          {
            "uri": "https://covers.openlibrary.org/b/isbn/978-0553103540-M.jpg",
          }
        }
        style={
          {
            "borderRadius": 8,
            "objectFit": "cover",
          }
        }
        testID="img978-0553103540"
        width={250}
      />
      <View
        style={
          {
            "backgroundColor": "rgba(0,0,0,0.5)",
            "borderBottomLeftRadius": 8,
            "borderBottomRightRadius": 8,
            "bottom": 0,
            "height": 30,
            "justifyContent": "center",
            "left": 0,
            "paddingHorizontal": 8,
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            [
              {
                "color": "black",
                "fontFamily": "System",
              },
              {
                "color": "#F7F7F7",
                "fontSize": 12,
                "fontWeight": "bold",
              },
              undefined,
            ]
          }
        >
          A Game of Thrones
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "position": "absolute",
        "right": 8,
        "top": 8,
      }
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#FFF400",
            "fontSize": 24,
          },
          undefined,
          {
            "fontFamily": "anticon",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
      testID="favoriteIcon"
    >
      î™ 
    </Text>
  </View>
</View>
`;
